flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java

```diff
@@ -18,35 +18,44 @@
 
 package org.apache.flink.runtime.execution.librarycache;
 
+import org.apache.flink.annotation.VisibleForTesting;
 import org.apache.flink.util.ChildFirstClassLoader;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
 import java.net.URL;
 import java.net.URLClassLoader;
+import java.util.Enumeration;
 
 /**
  * Gives the URLClassLoader a nicer name for debugging purposes.
  */
 public class FlinkUserCodeClassLoaders {
 
-	public static URLClassLoader parentFirst(URL[] urls, ClassLoader parent) {
+	@VisibleForTesting
+	static URLClassLoader parentFirst(URL[] urls, ClassLoader parent) {
```

Comment:
I would not reduce the visibility here, that is unrelated to the change, and I actually know users that have used this,

Reply:
Hm it's not public API and they could achieve the same thing through `#create` with one extra parameter. 
Unless I change the signature of this method, this method cannot use the safety net (or safety net would also need to be URLClassLoader which I wanted to avoid), which may also be fine, but I wanted to point that out explicitly.

Reply:
Why do you not want the safety net to be a URLClassLoader?

Reply:
I was not considering it initially because delegation pattern on implementation level felt wrong, but it's messy for ClassLoader anyways and it would solve this discussion, so I adjusted the PR.
Now we just need to converge on `loadClass`.

Reply:
Fair enough